File: baekJoon/17298.py
from sys import stdin

listSize = int(stdin.readline())
numList = list(map(int, stdin.readline().split()))
stack = [numList.pop()]
ans = [-1]


def getNum(stack, k):
    lenStack = len(stack)
    for i in range(lenStack-1, -1, -1):
        if stack[i] > k:
            return stack[i]
    return -1 
    

while len(numList)> 0:
    tmp = numList.pop()
    ans.append(getNum(stack, tmp))
    stack.append(tmp)
ans = ans[::-1]
    

for i in range(listSize-1):
    print(ans[i], end=" ")
print(ans[listSize-1])
    
    

File: baekJoon/17298.cpp
#include <iostream>
#include <stack>
#include <stdlib.h>

using namespace std;

int main() {
    ios_base::sync_with_stdio(false);

    stack<int> s;
    int ans = 0, size, tmp, m=0;

    cin >> size;
    int* arr = (int *)malloc(sizeof(int) * (size+1));
    int* ansArr = (int *)malloc(sizeof(int) * size); 
    
    for(int i=0; i<size; i++) {
        cin >> arr[i];
    }
        
    for(int i=size-1; i>=0; i--){
        if(arr[i] >= m){
            ansArr[i] = -1;
            m = arr[i];
            s.push(arr[i]);
            continue;
        }
        while(s.top() <= arr[i]){
            s.pop();
        }

        ansArr[i] = s.top();
        s.push(arr[i]);

    }

    for(int i=0; i<size; i++)
        cout << ansArr[i] << ' ';
    cout <<'\n';
    return 0;
}